### PR TITLE
show informative error when fetchers fail

### PR DESCRIPTION
**Pull request by haampie**
not sure why this was not done years ago...

### CODE REVIEW

**Review by alalazo — DISMISSED** [review]

**Comment by white238** [comment]
@kennyweiss @becker33 

Not sure if this branch made my error better but it did fix downloading. Was getting this on develop since yesterday:

```
==> No binary for vtk-m-2.2.0-xwolac5e3325cxvlfyowytgsxmbsirst found: installing from source
==> Error: timeout: The read operation timed out
```

With this branch it successfully downloaded and built.

**Comment by haampie** [comment]
> Not sure if this branch made my error better but it did fix downloading.

Nope ;) another pr #45683 merged yesterday made this error you had non-fatal so spack would continue if there was another mirror to fetch from -- but there isn't. This PR just presents errors in a better way (what urls were tried & what went wrong). The issue with vtk-m is really a server issue: sometimes a download gets stuck 50MB of downloading in.

File: lib/spack/spack/util/web.py [hunk 1]
```diff
@@ -307,45 +308,44 @@ def base_curl_fetch_args(url, timeout=0):
...
 
 
-def _curl(curl=None):
-    if not curl:
-        try:
-            curl = which("curl", required=True)
-        except CommandNotFoundError as exc:
-            tty.error(str(exc))
-            raise spack.error.FetchError("Missing required curl fetch method")
+def require_curl() -> Executable:
```
Comment (alalazo) [inline lib/spack/spack/util/web.py:338]: :+1: for removing the argument

File: lib/spack/spack/fetch_strategy.py [hunk 2]
```diff
@@ -1366,17 +1367,23 @@ def fetch(self):
         if parsed_url.scheme != "gs":
             raise spack.error.FetchError("GCSFetchStrategy can only fetch from gs:// urls.")
 
-        tty.debug("Fetching {0}".format(self.url))
-
         basename = os.path.basename(parsed_url.path)
+        request = urllib.request.Request(
+            self.url, headers={"User-Agent": web_util.SPACK_USER_AGENT}
+        )
```
Comment (alalazo) [inline lib/spack/spack/fetch_strategy.py:1373]: (minor) In a following PR we can put this in a function, to attach the headers that gets repeated.

**Review by alalazo — APPROVED** [review]

**Pull request merged**